godel_scan_analysis/src/keyence_scan_server.cpp

```diff
@@ -0,0 +1,79 @@
+#include "godel_scan_analysis/keyence_scan_server.h"
+
+#include <pcl_ros/transforms.h>
+#include <pcl/filters/voxel_grid.h>
+
+/*
+*/
+
+godel_scan_analysis::ScanServer::ScanServer(const std::string& world_frame, 
+                                            const std::string& scan_frame)
+  : map_(new ColorCloud)
+  , buffer_(new ColorCloud)
+  , from_frame_(world_frame)
+  , to_frame_(scan_frame)
+{
+  ros::NodeHandle nh;
+  scan_sub_ = nh.subscribe("profiles", 500, &ScanServer::scanCallback, this);
+  cloud_pub_ = nh.advertise<ColorCloud>("color_cloud", 1);
+  map_->header.frame_id = from_frame_;
+
+  // Create publisher for the collected color cloud
+  timer_ = nh.createTimer(ros::Duration(2.0), &ScanServer::publishCloud, this);
+}
+
+void godel_scan_analysis::ScanServer::scanCallback(const Cloud& cloud)
+{
+  // Generate colored point cloud of scan data
+  if (!scorer_.analyze(cloud, *buffer_)) return;
+
+  // Calculate time stamp was processed
+  ros::Time stamp;
+  stamp.fromNSec(cloud.header.stamp * 1000);
+
+  try
+  {
+    // Transform scan from optical frame to world frame
+    transformScan(*buffer_, stamp);    
+  }
+  catch(...)
+  {
+    ROS_WARN("No TF");
+    return;
+  }
+
+  // Insert into results cloud
+  map_->insert(map_->end(), buffer_->begin(), buffer_->end());
```

Comment:
This just keeps on buffering for as long as it's being produced. The Keyence can produce a lot of data so this can become an issue if left to run autonomously for a while. The resulting data is only really used for visual purposes at the moment, but will in the future be used to do automated re-planning on out-of-spec regions. For this reason, I believe that there should be some sort of `job` or `action` interface around this where data is stored to disk or deleted after the completion of an individual scan.